internal/jujuclient/dial.go

```diff
@@ -54,36 +56,101 @@ type Dialer struct {
 	JWTService                 *jimmjwx.JWTService
 }
 
-func (d *Dialer) createLoginRequest(ctx context.Context, ctl *dbmodel.Controller, modelTag names.ModelTag, p map[string]string) (*jujuparams.LoginRequest, error) {
-	// JIMM is automatically given all required permissions
-	permissions := p
-	if permissions == nil {
-		permissions = make(map[string]string)
-	}
-	permissions[ctl.ResourceTag().String()] = "superuser"
-	if modelTag.Id() != "" {
-		permissions[modelTag.String()] = "admin"
+func (d *Dialer) createLoginRequest(ctx context.Context, user *openfga.User, ctl *dbmodel.Controller, modelTag names.ModelTag, requirePermissionChecks map[string]string) (*jujuparams.LoginRequest, error) {
+
+	permissions := make(map[string]string)
+	if user == nil {
+		// if the user is not specify, we log in as the "admin" user
+		user = &openfga.User{
+			Identity: &dbmodel.Identity{
+				Name: "admin",
+			},
+		}
+
+		// if logging in as admin, we automatically grant the admin
+		// user all required permissions
+		if requirePermissionChecks != nil {
+			permissions = requirePermissionChecks
+		}
+
+		// we also state that the "admin" user is
+		// - controller superuser
+		// - model admin
+		permissions[ctl.ResourceTag().String()] = "superuser"
+		if modelTag.Id() != "" {
+			permissions[modelTag.String()] = "admin"
+		}
+	} else {
+		// if we are logging in as any other user, we need to check permissions
+
+		// controller permissions
+		switch user.GetControllerAccess(ctx, ctl.ResourceTag()) {
+		case ofganames.AdministratorRelation:
+			permissions[ctl.ResourceTag().String()] = "superuser"
+		default:
+			permissions[ctl.ResourceTag().String()] = "login"
+		}
+
+		if modelTag.Id() != "" {
+			// model permissions
+			switch user.GetModelAccess(ctx, modelTag) {
```

Comment:
What do you think about simplifying to just checking for `write` and `read`? Admin access is only needed in Juju for granting other users with access so we shouldn't need the admin's access (although maybe deleting a model requires admin? Not sure).